tokio/src/runtime/thread_id.rs

```diff
@@ -0,0 +1,46 @@
+use std::num::NonZeroU64;
+
+#[derive(Eq, PartialEq, Clone, Copy, Hash, Debug)]
+pub(crate) struct ThreadId(NonZeroU64);
+
+impl ThreadId {
+    cfg_has_atomic_u64! {
+        pub(crate) fn new() -> ThreadId {
+            use std::sync::atomic::{AtomicU64, Ordering::Relaxed};
+
+            static COUNTER: AtomicU64 = AtomicU64::new(0);
+
+            let mut last = COUNTER.load(Relaxed);
+            loop {
+                let id = match last.checked_add(1) {
+                    Some(id) => id,
+                    None => exhausted(),
+                };
+
+                match COUNTER.compare_exchange_weak(last, id, Relaxed, Relaxed) {
+                    Ok(_) => return ThreadId(NonZeroU64::new(id).unwrap()),
+                    Err(id) => last = id,
+                }
+            }
+        }
+    }
```

Comment:
It seems to me that doing the same as in `list.rs` would be sufficient here.

https://github.com/tokio-rs/tokio/blob/fbae620f0bd6cfc8e1c4ae188b97cead0b9b6f90/tokio/src/runtime/task/list.rs#L17-L54

Reply:
We do use the thread ID for correctness on top of the debug checks. It seems riskier, in that case, to fall back to u32.

Reply:
IIRC `std`'s thread IDs use a mutex on 32-bit targets, but I'd have to double check to make sure...

Reply:
Where do we use it for correctness?

Reply:
https://github.com/tokio-rs/tokio/blob/master/tokio/src/task/local.rs#L952